src/shared/GitHub/GitHubOAuth2Client.cs

```diff
@@ -1,33 +1,72 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
 using System;
 using System.Net.Http;
+using Microsoft.Git.CredentialManager;
 using Microsoft.Git.CredentialManager.Authentication.OAuth;
 
 namespace GitHub
 {
     public class GitHubOAuth2Client : OAuth2Client
     {
-        private static readonly string ClientId = "0120e057bd645470c1ed";
-        private static readonly string ClientSecret = "18867509d956965542b521a529a79bb883344c90";
-        private static readonly Uri RedirectUri = new Uri("http://localhost/");
-
-        public GitHubOAuth2Client(HttpClient httpClient, Uri baseUri)
-            : base(httpClient, CreateEndpoints(baseUri), ClientId, RedirectUri, ClientSecret) { }
+        public GitHubOAuth2Client(HttpClient httpClient, ISettings settings, Uri baseUri)
+            : base(httpClient, CreateEndpoints(baseUri),
+                GetClientId(settings), GetRedirectUri(settings), GetClientSecret(settings)) { }
 
         private static OAuth2ServerEndpoints CreateEndpoints(Uri baseUri)
         {
-            Uri authEndpoint = new Uri(baseUri, "/login/oauth/authorize");
-            Uri tokenEndpoint = new Uri(baseUri, "/login/oauth/access_token");
+            Uri authEndpoint = new Uri(baseUri, GitHubConstants.OAuthAuthorizationEndpointRelativeUri);
+            Uri tokenEndpoint = new Uri(baseUri, GitHubConstants.OAuthTokenEndpointRelativeUri);
 
             Uri deviceAuthEndpoint = null;
             if (GitHubConstants.IsOAuthDeviceAuthSupported)
             {
-                deviceAuthEndpoint = new Uri(baseUri, "/login/oauth/authorize/device");
+                deviceAuthEndpoint = new Uri(baseUri, GitHubConstants.OAuthDeviceEndpointRelativeUri);
             }
 
             return new OAuth2ServerEndpoints(authEndpoint, tokenEndpoint)
             {
                 DeviceAuthorizationEndpoint = deviceAuthEndpoint
             };
         }
+
+        private static string GetClientId(ISettings settings)
+        {
+            if (settings.TryGetSetting(
+                GitHubConstants.EnvironmentVariables.OAuthClientId,
+                Constants.GitConfiguration.Credential.SectionName, GitHubConstants.GitConfiguration.Credential.OAuthClientId,
+                out string clientId))
+            {
+                return clientId;
+            }
+
+            return GitHubConstants.OAuthClientId;
+        }
+
+        private static Uri GetRedirectUri(ISettings settings)
```

Comment:
This would let me route the redirect to something besides localhost.  I can't express
whether this bothers me or whether I'm just being paranoid.  I tried to walk thru the
MITM diagrams from the other day, but I'm in over my head I fear.  It may not be a
problem, but I don't know that.

I'm wondering what you would use this env var for in testing -- if as I understand it,
the responses need to come back into the builtin GCM server to advance the state
machine.  It seems like this would route the redirect somewhere else and you wouldn't
get responses back in the GCM.  But again, I don't completely understand the flow.

Alternatively, would it be safer to have this env var be under a #if DEBUG ?

Reply:
Redirect URIs in OAuth2 are tied to the client ID. That is to say each registered application must define at least one redirect URI, and during the various endpoint calls the client must send a  redirect URI that is associated with that app.

For example GCM just uses `http://localhost/` as the only redirect URI (loopback hosts are special cased in that the port is lazily matched; otherwise port must `===` too).

Other applications such as GitHub CLI use `http://127.0.0.1/callback` as the redirect URI (it includes the path `/callback`).

Also note that `127.0.0.1 != localhost` textually, so these are considered different redirect URIs.

> This would let me route the redirect to something besides localhost.

In theory yes, in practice no since the `OAuth2SystemWebBrowser` component that we use in GCM (the component that launches the browser, and starts the HTTP listener for the redirect) mandates that the redirect URI `Uri::IsLoopback` [here](https://github.com/microsoft/Git-Credential-Manager-Core/blob/7efe85294f661cb42fb336873355749c3d3c5d37/src/shared/Microsoft.Git.CredentialManager/Authentication/OAuth/OAuth2SystemWebBrowser.cs#L47-L50) and [here](https://github.com/microsoft/Git-Credential-Manager-Core/blob/7efe85294f661cb42fb336873355749c3d3c5d37/src/shared/Microsoft.Git.CredentialManager/Authentication/OAuth/OAuth2SystemWebBrowser.cs#L64-L67).

So we won't end up gifting `http://evil-server.com` with a authorisation code because:
1. We won't even start the auth flow via the browser unless we have a local redirect URI,
2. The server would need to agree to send the code to the redirect URI.

The 'problem' that might occur however is redirecting to _another_ localhost application listening on with a different path prefix. This is 'fixed' with PKCE as mentioned and implemented by GCM previously, but that requires server support.